### PR TITLE
Stage specification

### PR DESCRIPTION
**Pull request by bobaaaaa**
## Specification

Based on the latest designs (version zero, Stand: 22.10.2020) we specified the SectionPage model for the requirements:

- recursive `Stage`
- highly customizable `Stage` 'designs' (bg_color, title size, etc.)
- fixed set of grid sizes
- variable `Stage` content (teaser, ads, more stages)
- highly customizable `Teaser` 'designs' (variants, flags, etc.)

## CMCF on speed

We could build the current frontpage with only:

- 3 `StageLayout` (last project: min. 10)
- 3 `TeaserLayout` (last project: min. 12)

For the frontend each layout means a custom template.

## Unresolved

- links (references) part of a `Stage`

### CODE REVIEW

File: stroeer/pages/section/v1/section_page.proto [hunk 1]
```diff
@@ -13,12 +13,67 @@ message SectionPage {
...
+  TeaserLayout teaser_layout = 3;
+}
+
+// wip: common lingo discussion planned
+enum TeaserLayout {
+  TEASER_LAYOUT_UNSPECIFIED = 0;
+  TEASER_LAYOUT_HERO = 1;
+  TEASER_LAYOUT_CLUSTER = 2;
+  TEASER_LAYOUT_STORY = 4;
+  TEASER_LAYOUT_BULLET_ITEM = 5;
```
Comment (bobaaaaa) [inline stroeer/pages/section/v1/section_page.proto:68]:
```suggestion
```

Delete this. These teasers are part of the custom stages

File: stroeer/pages/section/v1/section_page.proto [hunk 2]
```diff
@@ -13,12 +13,67 @@ message SectionPage {
...
+  // ["sponsoring_logo":"gruender_szene.png"]
+  map<string, string> fields = 2;
+  TeaserLayout teaser_layout = 3;
+}
+
+// wip: common lingo discussion planned
+enum TeaserLayout {
+  TEASER_LAYOUT_UNSPECIFIED = 0;
+  TEASER_LAYOUT_HERO = 1;
+  TEASER_LAYOUT_CLUSTER = 2;
```
Comment (bobaaaaa) [inline stroeer/pages/section/v1/section_page.proto:66]: Maybe delete and built with default teaser + variant

File: stroeer/pages/section/v1/section_page.proto [hunk 3]
```diff
@@ -13,12 +13,67 @@ message SectionPage {
...
+  TEASER_LAYOUT_STORY = 4;
+  TEASER_LAYOUT_BULLET_ITEM = 5;
+}
+
+enum StageItemWidth {
+  STAGE_ITEM_WIDTH_UNSPECIFIED = 0;
+  STAGE_ITEM_WIDTH_FULL = 1;
+  STAGE_ITEM_WIDTH_HALF = 2;
+  reserved 3, 5 to 7; // use indices like the actual denomitators
+  STAGE_ITEM_WIDTH_QUARTER = 4;
```
Comment (thisismana) [inline stroeer/pages/section/v1/section_page.proto:76]: moar verbosity plz (cardinality), `STAGE_ITEM_WIDTH_ONE_QUARTER` or `STAGE_ITEM_WIDTH_THREE_QUARTER`

File: stroeer/pages/section/v1/section_page.proto [hunk 4]
```diff
@@ -13,12 +13,67 @@ message SectionPage {
...
+}
+
+message StageItem {
+  oneof item_type {
+    Stage stage = 1;
+    ArticleTeaser article_teaser = 2;
+    Commercial commercial = 3;
+  }
+  reserved 4 to 10; // for future item types
+  StageItemWidth stage_item_width = 11;
```
Comment (thisismana) [inline stroeer/pages/section/v1/section_page.proto:48]: move this to index `1` and let the `oneof` occupy the other indices

File: stroeer/pages/section/v1/section_page.proto [hunk 5]
```diff
@@ -13,12 +13,67 @@ message SectionPage {
...
+}
+
+message Commercial {
+  map<string, string> fields = 1;
+}
+
+message ArticleTeaser {
+  stroeer.core.v1.Article article = 1;
+  // ["sponsoring_logo":"gruender_szene.png"]
```
Comment (thisismana) [inline stroeer/pages/section/v1/section_page.proto:57]: remove reference to Springer affiliated stuff?

File: stroeer/pages/section/v1/section_page.proto (hunk 4 above)
Comment (moritzzimmer) [reply stroeer/pages/section/v1/section_page.proto:48]: reserving seems to be meant for deleted fields: https://developers.google.com/protocol-buffers/docs/proto3#reserved

File: stroeer/pages/section/v1/section_page.proto [hunk 6]
```diff
@@ -13,12 +13,67 @@ message SectionPage {
...
+  }
+  reserved 4 to 10; // for future item types
+  StageItemWidth stage_item_width = 11;
+}
+
+message Commercial {
+  map<string, string> fields = 1;
+}
+
+message ArticleTeaser {
```
Comment (thisismana) [inline stroeer/pages/section/v1/section_page.proto:55]: also use this in `ArticlePage.related`?

File: stroeer/pages/section/v1/section_page.proto (hunk 5 above)
Comment (harryurban) [reply stroeer/pages/section/v1/section_page.proto:57]: this is just a temporary comment
all comments, that should not be part of the generated documentation, should be deleted before merging this PR

File: stroeer/pages/section/v1/section_page.proto (hunk 2 above)
Comment (harryurban) [reply stroeer/pages/section/v1/section_page.proto:66]: but then we have an enum without any values ;)
should we still keep the Enum?

File: stroeer/pages/section/v1/section_page.proto (hunk 6 above)
Comment (bobaaaaa) [reply stroeer/pages/section/v1/section_page.proto:55]: I like the idea. Let's check this in a separate PR 👍

**Pull request merged**